[FIX JENKINS-51495] descriptorRadioList does not honor DescriptorVisibilityFilter

matsushou opened the pull request:
See [JENKINS-51495](https://issues.jenkins-ci.org/browse/JENKINS-51495).

### Proposed changelog entries

Oleg's version:
* Update descriptorRadioList form elements to honor DescriptorVisibilityFilter extension points

Original version
*  Quit using index to associate descriptor selected on the front-end with the implementation on the back-end, and begin using descriptor id instead.

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

matsushou committed:
[FIX JENKINS-51495] descriptorRadioList does not honor DescriptorVisibilityFilter

quit using index to associate descriptor selected on the front-end with the implementation on the back-end.

batmat commented:
Test failure seems real?

matsushou committed:
[FIX JENKINS-51495] descriptorRadioList does not honor DescriptorVisibilityFilter

fix additionally because of test failures.

matsushou commented:
Yes, it seems real, and all 5 failures are due to not updating DescriptorExtensionList...
I've added a similar commit.

matsushou commented:
@batmat 

Could you review this PR again (or remove needs-fix label in order others can review) ?

oleg-nenashev reviewed (approved):
LGTM, thanks @matsushou !

fcojfernandez reviewed (approved):
LGTM

oleg-nenashev commented:
i will land it into the next weekly if there is no negative feedback

daniel-beck commented:
Would like to see a test, because `DescriptorVisibilityFilter` is not something regularly encountered AFAICT.

matsushou commented:
Jenkins failed because of the error below.

`Failed to install https://download.java.net/java/GA/jdk11/9/GPL/openjdk-11.0.2_windows-x64_bin.zip to C:\Jenkins\tools\hudson.model.JDK\jdk11`

For me, it seems to be an environmental problem, not sure though...
I guess re-run may work.

oleg-nenashev commented:
https://issues.jenkins-ci.org/browse/INFRA-2086 FYI. The Oracle's service is still broken, not that much I can do about it without access to the infra.

putting on hold so that you can discuss @daniel-beck 's comment

daniel-beck commented:
@matsushou Only affects the Windows PR builds, please ignore them for now.

matsushou commented:
> Only affects the Windows PR builds, please ignore them for now.

Thank you for your comment.
At last, is this PR assumed to be ready-for-merge again?

oleg-nenashev commented:
IIUC https://github.com/jenkinsci/jenkins/pull/3969#issuecomment-487283884 is still open @matsushou

matsushou commented:
sorry, I misunderstood the meaning of the comment...
I'll add some test cases shortly.

oleg-nenashev commented:
@matsushou please let us know if you need any assistance

matsushou commented:
thank you so much, I'll take time to solve it this weekend at the latest.

matsushou committed:
[FIX JENKINS-51495] descriptorRadioList does not honor
DescriptorVisilityFilter

add a test method for DescriptorList and DescriptorExtensionList.

matsushou commented:
I added a test case but couldn't run it locally... (need to wait Jenkins build)
Also, I'm not sure it's enough or an appropriate way to write a test in this case...

matsushou committed:
[FIX JENKINS-51495] descriptorRadioList does not honor DescriptorVisibilityFilter

fix test failure because of the reuse of Map.

matsushou committed:
[FIX JENKINS-51495] descriptorRadioList does not honor DescriptorVisibilityFilter

fix test failure by changing class to find.

matsushou commented:
began to fail without additional push... ❓ 

https://ci.jenkins.io/blue/organizations/jenkins/Core%2Fjenkins/activity?branch=PR-3969

oleg-nenashev reviewed (commented):
Thatks, and sorry for the review delay. I checked the code, and I am not sure that returning null from `newInstanceFromRadioList` is actually an expected behavior. Maybe it makes sense to just throw an error. Will double-check the code to provide a definite answer

oleg-nenashev commented:
Still om my side :(

daniel-beck commented:
There is another use case for this:

When installation of a plugin moves the indexes around. If the config form was loaded before the plugin's extensions were loaded, but submitted after, the index in the list will not match the type, and an error will be shown to the user.

MRamonLeon commented:
@oleg-nenashev how can we move it on? Are you going to look at it and provide an answer? thanks!

oleg-nenashev commented:
Sorry, I missed the comments. Please request m reviews when needed so that it appears on my dashboard

oleg-nenashev reviewed (commented):
I will submit a follow-up patch for it. Creating a null object creates a lot of potential problems down the line in usages, including unintentional security settings disabling if the property is filtered.

I will create a follow-up the request to suggest my version

alecharp commented:
Closing to trigger a new build on Jenkins.

timja commented:
> I will submit a follow-up patch for it. Creating a null object creates a lot of potential problems down the line in usages, including unintentional security settings disabling if the property is filtered.
> 
> I will create a follow-up the request to suggest my version

to be clear are you blocking this @oleg-nenashev? or happy for it to be merged?

oleg-nenashev commented:
Sorry, I forgot to act on this pull request. In my TODO list for this week

oleg-nenashev reviewed (approved):
I reviewed the code, and I confirm that the suggested change is safe.

* All usages: https://github.com/search?q=org%3Ajenkinsci+newInstanceFromRadioList&type=Code
* There are 3 usages in the Jenkins Core. All of them handle nulls properly. I will be submitting a follow-up pull request(s) with annotation fixes and one unrelated bug I discovered
* There is one usage in Groovy Plugin. https://github.com/jenkinsci/groovy-plugin/blob/6e423e98dd819baedbc15ec93410b1719e4cf574/src/main/java/hudson/plugins/groovy/AbstractGroovy.java#L35-L66 .  Metod - getScriptSource()
  * This usage is in the deprecated method, but apparently it is used in the Configuration Slicing plugin: https://github.com/jenkinsci/configurationslicing-plugin/blob/2fa473ed02142b0e97341f9b3b182ea8ac4c52cc/src/main/java/configurationslicing/tools/GroovySlicer.java#L46 . This is the functionality available to admins only
  * Configuration Slicer Plugin invokes Groovy execution. When it is ran, https://github.com/jenkinsci/groovy-plugin/blob/6e423e98dd819baedbc15ec93410b1719e4cf574/src/main/java/hudson/plugins/groovy/Groovy.java#L64-L67 will process `null` and fail the execution when the required descriptor is filtered by DescriptorVisibilityFilter 
  * Although this behavior is not ideal, IMO it does not constitute to a horrible regression. Admins will be able to track down the problem if needed, and I would argue that it is better to fail the execution if it is filtered on the instance
* There is one usage in the proprietary plugins I have access to. This usage handles nulls properly

res0nance reviewed (commented):
Looks ok, https://github.com/jenkinsci/jenkins/pull/3969#issuecomment-487283884 still seems open though.

oleg-nenashev committed:
[JENKINS-51495] - Fix formatting of the unit test

oleg-nenashev commented on test/src/test/java/hudson/ExtensionListTest.java:
```diff
@@ -174,6 +177,28 @@ public void legacyDescriptorList() throws Exception {
 
+	@Test
+	public void newInstanceFromRadioList() throws Exception {
+		// test for DescriptorList
```
Fixed, thanks!

oleg-nenashev commented:
I plan to merge it in 24 hours if no negative feedback

oleg-nenashev commented:
Thanks a lot @matsushou ! Sorry that it took so much time to get it merged
